doc: clarify extensionless esm (#30657)
nodejs/node: doc/api/esm.md

`"type": "commonjs"`, extensionless and `.js` files are treated as CommonJS.
If the volume root is reached and no `package.json` is found,
Node.js defers to the default, a `package.json` with no `"type"`

> field. Note that "extensionless" refers to file names which do not contain

Reviewer: ```suggestion
field. Note that "extensionless" refers to file paths which do not contain
```

> field. Note that "extensionless" refers to file paths which do not contain

Reviewer: ```suggestion
field. "Extensionless" refers to file paths which do not contain
```

Our Doc linter bans the use of `note that`

> an extension. It does not refer to the import specifier.

Reviewer: ```suggestion
an extension as opposed to optionally dropping a file extension in a specifier.
```

```js
// my-app.js, part of the same example as above
import './startup.js'; // Loaded as ES module because of package.json

// Extensionless. Requires 'node --es-module-specifier-resolution=node'

> import './startup'; // Loaded as ES module because of package.json

Reviewer: Extensionless files, as documented above, are different from extensionless specifiers (which you are documenting here). Extensionless files mean the file path itself has no extension e.g. `/path/to/file`, not the specifier e.g. `'./file'` where path is `/path/to/file.js`.

Author: Ah I see. Yeah it could definitely be clarified.

Author: Done. Had trouble writing it succinctly so let me know if it makes sense.
